<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>coFAST: NSCLC CosMx data coembedding • coFAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="coFAST: NSCLC CosMx data coembedding">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coFAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CosMx.html">coFAST: NSCLC CosMx data coembedding</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/coFAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>coFAST: NSCLC CosMx data coembedding</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2025-03-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/coFAST/blob/HEAD/vignettes/CosMx.Rmd" class="external-link"><code>vignettes/CosMx.Rmd</code></a></small>
      <div class="hidden name"><code>CosMx.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the coFAST workflow for the analysis of
NSCLC CosMx spatial transcriptomics dataset. In this vignette, the
workflow of coFAST consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Coembedding dimension reduction</li>
<li>Downstream analysis (i.e. , signature gene analysis, visualization
of cell types and coembeddings,)</li>
</ul>
<div class="section level2">
<h2 id="load-and-view-data">Load and view data<a class="anchor" aria-label="anchor" href="#load-and-view-data"></a>
</h2>
<p>We demonstrate the use of coFAST to NSCLC data, which can be
downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2024</span><span class="op">)</span> <span class="co"># set a random seed for reproducibility.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/coFAST" class="external-link">coFAST</a></span><span class="op">)</span> <span class="co"># load the package of coFAST method</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span>
<span><span class="va">CosMx_subset</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 960 features across 2814 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (960 features, 960 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<p>First, we normalize the data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span></code></pre></div>
<p>Then, we select the variable genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="coembedding-using-cofast">Coembedding using coFAST<a class="anchor" aria-label="anchor" href="#coembedding-using-cofast"></a>
</h2>
<p>We introduce how to use coFAST to perform coembedding for this CosMx
data. First, we determine the dimension of coembeddings. Then, we select
the variable genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diagnostic.cor.eigs.html">diagnostic.cor.eigs</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_est</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">dat_cor</span>, <span class="st">"q_est"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"q_est = "</span>, <span class="va">q_est</span>, <span class="st">'\n'</span><span class="op">)</span></span>
<span><span class="co">#&gt; q_est =  28</span></span></code></pre></div>
<p>Subsequently, we calculate coembeddings by utilizing coFAST, and
observe that the <code>reductions</code> field acquires an additional
component named <code>fast</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="co"># Extract the spatial coordinates</span></span>
<span><span class="va">Adj_sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdj.html">AddAdj</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span> <span class="co">## calculate the adjacency matrix</span></span>
<span><span class="co">#&gt; The spatial cooridnates are 2 dimensions</span></span>
<span><span class="co">#&gt; Find the adjacency matrix by bisection method...</span></span>
<span><span class="co">#&gt; Current radius is 62.34</span></span>
<span><span class="co">#&gt; Median of neighborhoods is 0</span></span>
<span><span class="co">#&gt; Current radius is 93.01</span></span>
<span><span class="co">#&gt; Median of neighborhoods is 2</span></span>
<span><span class="co">#&gt; Current radius is 93.01</span></span>
<span><span class="co">#&gt; Median of neighborhoods is 5</span></span>
<span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coFAST.html">coFAST</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, Adj_sp <span class="op">=</span> <span class="va">Adj_sp</span>, q <span class="op">=</span> <span class="va">q_est</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finish variable intialization </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 2, loglik= -1426036.512228, dloglik=0.999336 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 3, loglik= -1412722.734084, dloglik=0.009336 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 4, loglik= -1409935.716061, dloglik=0.001973 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 5, loglik= -1408112.792632, dloglik=0.001293 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 6, loglik= -1407078.332386, dloglik=0.000735 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 7, loglik= -1406307.821844, dloglik=0.000548 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 8, loglik= -1405752.638142, dloglik=0.000395 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 9, loglik= -1405322.991573, dloglik=0.000306 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 10, loglik= -1404988.225441, dloglik=0.000238 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 11, loglik= -1404720.086242, dloglik=0.000191 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 12, loglik= -1404502.801026, dloglik=0.000155 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 13, loglik= -1404323.837749, dloglik=0.000127 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 14, loglik= -1404174.551626, dloglik=0.000106 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 15, loglik= -1404048.246092, dloglik=0.000090 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 16, loglik= -1403940.000384, dloglik=0.000077 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 17, loglik= -1403846.063218, dloglik=0.000067 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 18, loglik= -1403763.627389, dloglik=0.000059 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 19, loglik= -1403690.551299, dloglik=0.000052 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 20, loglik= -1403625.203865, dloglik=0.000047 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 21, loglik= -1403566.324126, dloglik=0.000042 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 22, loglik= -1403512.932206, dloglik=0.000038 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 23, loglik= -1403464.256781, dloglik=0.000035 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 24, loglik= -1403419.685684, dloglik=0.000032 </span></span>
<span><span class="co">#&gt; Satrt ICM and E-step! </span></span>
<span><span class="co">#&gt; Finish ICM and E-step! </span></span>
<span><span class="co">#&gt; iter = 25, loglik= -1403378.726805, dloglik=0.000029</span></span>
<span><span class="co">#&gt; 2025-03-15 20:08:02.522399 : ***** Finish CoFAST, 0.129 mins elapsed.</span></span>
<span><span class="va">CosMx_subset</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 960 features across 2814 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (960 features, 960 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  1 dimensional reduction calculated: cofast</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downstream-analysis">Downstream analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h2>
<div class="section level3">
<h3 id="clustering-and-assess-the-cluster-significance">Clustering and assess the cluster significance<a class="anchor" aria-label="anchor" href="#clustering-and-assess-the-cluster-significance"></a>
</h3>
<p>First, we show how to use the spot embedding for spatial domain
segmentation. User can set the number of clusters as fixed value, i.e.,
K=10; or set K=NULL using the default resolution. The cluster results
are in the <code>cofast.cluster</code> column of data.frame
<code>meta.data</code> slot. The Idents of Seurat object is also set as
<code>cofast.cluster</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddCluster.html">AddCluster</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 2814</span></span>
<span><span class="co">#&gt; Number of edges: 105139</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8115</span></span>
<span><span class="co">#&gt; Number of communities: 12</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          orig.ident nCount_RNA nFeature_RNA        x        y   cell_type</span></span>
<span><span class="co">#&gt; c_1_1_2           c         26           23 6092.889 158834.7 endothelial</span></span>
<span><span class="co">#&gt; c_1_1_3           c         74           51 7214.889 158843.7 endothelial</span></span>
<span><span class="co">#&gt; c_1_1_4           c         60           48 7418.889 158813.7         pDC</span></span>
<span><span class="co">#&gt; c_1_1_5           c         52           39 7446.889 158845.7  macrophage</span></span>
<span><span class="co">#&gt; c_1_1_7           c        419          153 3902.889 158820.7  macrophage</span></span>
<span><span class="co">#&gt; c_1_1_8           c         42           35 3992.889 158832.7 endothelial</span></span>
<span><span class="co">#&gt; c_1_1_9           c        249          103 4067.889 158827.7 plasmablast</span></span>
<span><span class="co">#&gt; c_1_1_10          c         99           52 4111.889 158842.7         pDC</span></span>
<span><span class="co">#&gt; c_1_1_11          c         98           62 4254.889 158841.7         mDC</span></span>
<span><span class="co">#&gt; c_1_1_12          c        173           90 4495.889 158830.7  macrophage</span></span>
<span><span class="co">#&gt;          cofast.cluster seurat_clusters</span></span>
<span><span class="co">#&gt; c_1_1_2               6               5</span></span>
<span><span class="co">#&gt; c_1_1_3               6               5</span></span>
<span><span class="co">#&gt; c_1_1_4               3               2</span></span>
<span><span class="co">#&gt; c_1_1_5               5               4</span></span>
<span><span class="co">#&gt; c_1_1_7               5               4</span></span>
<span><span class="co">#&gt; c_1_1_8               6               5</span></span>
<span><span class="co">#&gt; c_1_1_9               7               6</span></span>
<span><span class="co">#&gt; c_1_1_10              7               6</span></span>
<span><span class="co">#&gt; c_1_1_11              5               4</span></span>
<span><span class="co">#&gt; c_1_1_12              5               4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1   2   3   4   5   6   7   8   9  10  11  12 </span></span>
<span><span class="co">#&gt; 402 398 396 333 317 276 226 151 109  89  72  45</span></span></code></pre></div>
<p>We visualize the clusters on spatial coordinates.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Addcoord2embed.html">Addcoord2embed</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, coord.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 960 features across 2814 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (960 features, 960 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: cofast, Spatial</span></span>
<span><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu">PRECAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/PRECAST/man/chooseColors.html" class="external-link">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">'Classic 20'</span>, n_colors<span class="op">=</span><span class="fl">20</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DimPlot(CosMx_subset, reduction='Spatial', cols=cols_cluster, pt.size = 2)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction<span class="op">=</span><span class="st">'Spatial'</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cofast.cluster"</span>, <span class="st">'cell_type'</span><span class="op">)</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, pt.size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-8-2.png" width="960"></p>
<p>Next, we assess the aggregation scores of each cluster on spatial
space and embedding space. We calculated the spatial aggregation scores
and found that cluster 8 (corresponding to tumor) has the highest
spatial aggregation score, which is concordance with the spatial
clustering map.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat.spa.score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AggregationScore.html">AggregationScore</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction.name <span class="op">=</span> <span class="st">'Spatial'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ade4</span></span>
<span><span class="co">#&gt; Warning: package 'ade4' was built under R version 4.4.2</span></span>
<span><span class="co">#&gt; Warning: package 'furrr' was built under R version 4.4.2</span></span>
<span><span class="co">#&gt; Loading required package: future</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dat.spa.score</span><span class="op">)</span></span>
<span><span class="co">#&gt;    num.spots  AggScore</span></span>
<span><span class="co">#&gt; 1        402 10.849475</span></span>
<span><span class="co">#&gt; 2        398  9.867601</span></span>
<span><span class="co">#&gt; 3        396 33.444585</span></span>
<span><span class="co">#&gt; 4        333 24.748565</span></span>
<span><span class="co">#&gt; 5        317 16.587576</span></span>
<span><span class="co">#&gt; 6        276 10.080736</span></span>
<span><span class="co">#&gt; 7        226 19.048832</span></span>
<span><span class="co">#&gt; 8        151 99.386092</span></span>
<span><span class="co">#&gt; 9        109 27.294890</span></span>
<span><span class="co">#&gt; 10        89 16.201003</span></span>
<span><span class="co">#&gt; 11        72  3.558083</span></span>
<span><span class="co">#&gt; 12        45  2.638801</span></span></code></pre></div>
<p>Next, we calculated the embedding aggregation scores and found that
cluster 8 also has the highest embedding aggregation score.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat.embd.score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AggregationScore.html">AggregationScore</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction.name <span class="op">=</span> <span class="st">'cofast'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dat.embd.score</span><span class="op">)</span></span>
<span><span class="co">#&gt;    num.spots  AggScore</span></span>
<span><span class="co">#&gt; 1        402  70.74596</span></span>
<span><span class="co">#&gt; 2        398  82.71345</span></span>
<span><span class="co">#&gt; 3        396  71.30350</span></span>
<span><span class="co">#&gt; 4        333  71.43045</span></span>
<span><span class="co">#&gt; 5        317  83.63599</span></span>
<span><span class="co">#&gt; 6        276  64.88092</span></span>
<span><span class="co">#&gt; 7        226  78.59129</span></span>
<span><span class="co">#&gt; 8        151 108.42777</span></span>
<span><span class="co">#&gt; 9        109  89.64092</span></span>
<span><span class="co">#&gt; 10        89  84.97524</span></span>
<span><span class="co">#&gt; 11        72  84.46171</span></span>
<span><span class="co">#&gt; 12        45  76.56332</span></span></code></pre></div>
<p>In the following, we show how to find the signature genes based on
comebeddings. First, we calculate the distance matrix.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pdistance.html">pdistance</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"cofast"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculate co-embedding distance...</span></span></code></pre></div>
<p>Next, we find the signature genes for each cell type</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1   2   3   4   5   6   7   8   9  10  11  12 </span></span>
<span><span class="co">#&gt; 402 398 396 333 317 276 226 151 109  89  72  45</span></span>
<span><span class="co">#Idents(CosMx_subset) &lt;- CosMx_subset$cell_type</span></span>
<span><span class="va">df_sig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find.signature.genes.html">find.signature.genes</a></span><span class="op">(</span><span class="va">CosMx_subset</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">df_sig_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 12</span></span>
<span><span class="co">#&gt;  $ 1 :'data.frame':  327 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:327] 17.2 17.3 17.3 17.4 17.4 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:327] 0.706 0.254 0.781 0.303 0.311 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:327] "1" "1" "1" "1" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:327] "COL1A2" "COL5A2" "FN1" "PDGFRA" ...</span></span>
<span><span class="co">#&gt;  $ 10:'data.frame':  207 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:207] 15.9 18.4 19.2 19.3 19.7 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:207] 0.865 0.315 0.225 0.213 0.348 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:207] "10" "10" "10" "10" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:207] "CXCL8" "IL1R2" "HCAR2" "S100A8" ...</span></span>
<span><span class="co">#&gt;  $ 11:'data.frame':  285 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:285] 16.1 16.3 16.9 17.4 18.3 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:285] 0.972 0.708 1 1 0.806 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:285] "11" "11" "11" "11" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:285] "CPA3" "IL1RL1" "TPSB2" "TPSAB1" ...</span></span>
<span><span class="co">#&gt;  $ 12:'data.frame':  337 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:337] 17.8 18.1 18.4 18.4 18.5 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:337] 0.933 0.711 0.311 0.733 0.511 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:337] "12" "12" "12" "12" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:337] "KRT7" "VEGFA" "KRT4" "CD9" ...</span></span>
<span><span class="co">#&gt;  $ 2 :'data.frame':  382 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:382] 15.2 15.3 15.5 15.8 15.8 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:382] 0.842 0.55 0.407 0.352 0.477 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:382] "2" "2" "2" "2" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:382] "VWF" "CDH5" "CLEC14A" "ADGRL2" ...</span></span>
<span><span class="co">#&gt;  $ 3 :'data.frame':  358 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:358] 16.3 16.5 16.5 16.6 16.6 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:358] 0.354 0.48 0.323 0.194 0.598 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:358] "3" "3" "3" "3" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:358] "CD52" "CD37" "MS4A1" "TNFRSF13B" ...</span></span>
<span><span class="co">#&gt;  $ 4 :'data.frame':  261 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:261] 15.2 15.3 15.3 15.6 15.6 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:261] 0.27 0.321 0.465 0.441 0.246 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:261] "4" "4" "4" "4" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:261] "ITK" "CD2" "IL7R" "PTPRC" ...</span></span>
<span><span class="co">#&gt;  $ 5 :'data.frame':  326 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:326] 16.5 16.6 16.7 16.9 17.1 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:326] 0.716 0.808 0.558 0.306 0.836 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:326] "5" "5" "5" "5" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:326] "CD163" "C1QB" "GPNMB" "MS4A4A" ...</span></span>
<span><span class="co">#&gt;  $ 6 :'data.frame':  76 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:76] 18.4 18.4 18.6 18.8 18.8 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:76] 0.826 0.272 0.656 0.634 0.558 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:76] "6" "6" "6" "6" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:76] "MZT2A" "UBE2C" "DUSP5" "TYK2" ...</span></span>
<span><span class="co">#&gt;  $ 7 :'data.frame':  260 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:260] 14.1 14.8 15.4 15.7 15.8 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:260] 0.743 0.717 0.354 0.664 0.903 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:260] "7" "7" "7" "7" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:260] "MZB1" "JCHAIN" "CD79A" "IGHG2" ...</span></span>
<span><span class="co">#&gt;  $ 8 :'data.frame':  412 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:412] 15.5 16.4 17.7 18.3 18.3 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:412] 0.781 0.934 0.795 0.742 0.781 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:412] "8" "8" "8" "8" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:412] "PSCA" "CEACAM6" "CXCL5" "TACSTD2" ...</span></span>
<span><span class="co">#&gt;  $ 9 :'data.frame':  427 obs. of  4 variables:</span></span>
<span><span class="co">#&gt;   ..$ distance : num [1:427] 17.8 17.9 18.3 18.3 19.2 ...</span></span>
<span><span class="co">#&gt;   ..$ expr.prop: num [1:427] 0.798 0.798 0.963 0.642 0.514 ...</span></span>
<span><span class="co">#&gt;   ..$ label    : chr [1:427] "9" "9" "9" "9" ...</span></span>
<span><span class="co">#&gt;   ..$ gene     : chr [1:427] "LAMP3" "DMBT1" "LTF" "ITGB6" ...</span></span></code></pre></div>
<p>Then, we obtain the top five signature genes and organize them into a
data.frame. Next, we calculate the UMAP projections of coembeddings. The
colname <code>distance</code> means the distance between gene (i.e.,
IL7R) and cells with the specific cluster (i.e., cluster 1), which is
calculated based on the coembedding of genes and cells in the
coembedding space. The distance is smaller, the association between gene
and the cell type is stronger. The colname <code>expr.prop</code>
represents the expression proportion of the gene (i.e., IL7R) within the
cell type (i.e., tumor). The colname <code>label</code> means the
cluster and colname <code>gene</code> denotes the gene name. By the
data.frame object, we know <code>IL7R</code> is the one of the top
signature gene of cluster 1.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get.top.signature.dat.html">get.top.signature.dat</a></span><span class="op">(</span><span class="va">df_sig_list</span>, ntop <span class="op">=</span> <span class="fl">2</span>, expr.prop.cutoff <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span>
<span><span class="co">#&gt;        distance expr.prop label   gene</span></span>
<span><span class="co">#&gt; COL1A2 17.16143 0.7064677     1 COL1A2</span></span>
<span><span class="co">#&gt; COL5A2 17.25599 0.2537313     1 COL5A2</span></span>
<span><span class="co">#&gt; CXCL8  15.88425 0.8651685    10  CXCL8</span></span>
<span><span class="co">#&gt; IL1R2  18.42553 0.3146067    10  IL1R2</span></span>
<span><span class="co">#&gt; CPA3   16.13164 0.9722222    11   CPA3</span></span>
<span><span class="co">#&gt; IL1RL1 16.26874 0.7083333    11 IL1RL1</span></span></code></pre></div>
<p>Next, we calculate the UMAP projections of coembeddings of cells and
the selected signature genes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CosMx_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembedding_umap.html">coembedding_umap</a></span><span class="op">(</span></span>
<span>  <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"cofast"</span>, reduction.name <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  gene.set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Furthermore, we visualize the cells and top two signature genes of
tumor in the UMAP space of coembedding. We observe that the UMAP
projections of the two signature genes are near to tumor, which
indicates these genes are enriched in tumor.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## choose beutifual colors</span></span>
<span><span class="va">cols_cluster2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="va">cols_cluster</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span></span>
<span>   <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>   gene_txtdata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">dat</span>, <span class="va">label</span><span class="op">==</span><span class="st">'8'</span><span class="op">)</span>,</span>
<span>   cols<span class="op">=</span><span class="va">cols_cluster2</span>, pt_text_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>Then, we visualize the cells and top two signature genes of all
involved cell types in the UMAP space of coembedding. We observe that
the UMAP projections of the signature genes are near to the
corresponding cell type, which indicates these genes are enriched in the
corresponding cells.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coembed_plot.html">coembed_plot</a></span><span class="op">(</span></span>
<span>   <span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>   gene_txtdata <span class="op">=</span> <span class="va">dat</span>, cols<span class="op">=</span><span class="va">cols_cluster2</span>, </span>
<span>   pt_text_size <span class="op">=</span> <span class="fl">3</span>, alpha<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="va">p2</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-16-1.png" width="864"></p>
<p>In addtion, we can fully take advantages of the visualization
functions in <code>Seurat</code> package for visualization. The
following is an example that visualizes the cell types on the UMAP
space.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span><span class="op">)</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-17-1.png" width="864"></p>
<p>Then, there is another example that we plot the first two signature
genes of Tumor 5 on UMAP space, in which we observed the high expression
in tumor in constrast to other cell types.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">CosMx_subset</span>, reduction <span class="op">=</span> <span class="st">'UMAP'</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PSCA"</span>, <span class="st">"CEACAM6"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="CosMx_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] furrr_0.3.1        future_1.34.0      ade4_1.7-22        purrr_1.0.2       </span></span>
<span><span class="co">#&gt; [5] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           coFAST_0.1.0      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.4                    matrixStats_1.4.1          </span></span>
<span><span class="co">#&gt;   [3] spatstat.sparse_3.1-0       httr_1.4.7                 </span></span>
<span><span class="co">#&gt;   [5] RColorBrewer_1.1-3          tools_4.4.1                </span></span>
<span><span class="co">#&gt;   [7] sctransform_0.4.1           backports_1.5.0            </span></span>
<span><span class="co">#&gt;   [9] utf8_1.2.4                  R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [11] lazyeval_0.2.2              uwot_0.2.2                 </span></span>
<span><span class="co">#&gt;  [13] withr_3.0.1                 prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt;  [15] gridExtra_2.3               progressr_0.14.0           </span></span>
<span><span class="co">#&gt;  [17] cli_3.6.3                   Biobase_2.64.0             </span></span>
<span><span class="co">#&gt;  [19] textshaping_0.4.0           spatstat.explore_3.3-2     </span></span>
<span><span class="co">#&gt;  [21] fastDummies_1.7.4           labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.9                  spatstat.data_3.1-2        </span></span>
<span><span class="co">#&gt;  [25] ggridges_0.5.6              pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;  [27] pkgdown_2.1.1               systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt;  [29] harmony_1.2.1               scater_1.32.1              </span></span>
<span><span class="co">#&gt;  [31] parallelly_1.38.0           rstudioapi_0.16.0          </span></span>
<span><span class="co">#&gt;  [33] RSQLite_2.3.7               FNN_1.1.4.1                </span></span>
<span><span class="co">#&gt;  [35] generics_0.1.3              gtools_3.9.5               </span></span>
<span><span class="co">#&gt;  [37] ica_1.0-3                   spatstat.random_3.3-2      </span></span>
<span><span class="co">#&gt;  [39] car_3.1-3                   dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [41] Matrix_1.7-0                ggbeeswarm_0.7.2           </span></span>
<span><span class="co">#&gt;  [43] fansi_1.0.6                 S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [45] abind_1.4-8                 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt;  [47] yaml_2.3.10                 CompQuadForm_1.4.3         </span></span>
<span><span class="co">#&gt;  [49] carData_3.0-5               SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt;  [51] SparseArray_1.4.8           BiocFileCache_2.12.0       </span></span>
<span><span class="co">#&gt;  [53] Rtsne_0.17                  grid_4.4.1                 </span></span>
<span><span class="co">#&gt;  [55] blob_1.2.4                  promises_1.3.0             </span></span>
<span><span class="co">#&gt;  [57] crayon_1.5.3                GiRaF_1.0.1                </span></span>
<span><span class="co">#&gt;  [59] miniUI_0.1.1.1              lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [61] beachmat_2.20.0             cowplot_1.1.3              </span></span>
<span><span class="co">#&gt;  [63] KEGGREST_1.44.1             pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [65] knitr_1.48                  ProFAST_1.5                </span></span>
<span><span class="co">#&gt;  [67] GenomicRanges_1.56.2        future.apply_1.11.2        </span></span>
<span><span class="co">#&gt;  [69] codetools_0.2-20            leiden_0.4.3.1             </span></span>
<span><span class="co">#&gt;  [71] glue_1.7.0                  spatstat.univar_3.0-1      </span></span>
<span><span class="co">#&gt;  [73] data.table_1.16.0           vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [75] png_0.1-8                   spam_2.10-0                </span></span>
<span><span class="co">#&gt;  [77] org.Mm.eg.db_3.20.0         gtable_0.3.5               </span></span>
<span><span class="co">#&gt;  [79] cachem_1.1.0                xfun_0.47                  </span></span>
<span><span class="co">#&gt;  [81] S4Arrays_1.4.1              mime_0.12                  </span></span>
<span><span class="co">#&gt;  [83] survival_3.6-4              SingleCellExperiment_1.26.0</span></span>
<span><span class="co">#&gt;  [85] fitdistrplus_1.2-1          ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [87] nlme_3.1-164                bit64_4.5.2                </span></span>
<span><span class="co">#&gt;  [89] progress_1.2.3              filelock_1.0.3             </span></span>
<span><span class="co">#&gt;  [91] RcppAnnoy_0.0.22            GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [93] bslib_0.8.0                 irlba_2.3.5.1              </span></span>
<span><span class="co">#&gt;  [95] vipor_0.4.7                 KernSmooth_2.23-24         </span></span>
<span><span class="co">#&gt;  [97] colorspace_2.1-1            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt;  [99] DBI_1.2.3                   tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [101] DR.SC_3.4                   bit_4.5.0                  </span></span>
<span><span class="co">#&gt; [103] compiler_4.4.1              curl_5.2.3                 </span></span>
<span><span class="co">#&gt; [105] httr2_1.0.4                 BiocNeighbors_1.22.0       </span></span>
<span><span class="co">#&gt; [107] xml2_1.3.6                  desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [109] DelayedArray_0.30.1         plotly_4.10.4              </span></span>
<span><span class="co">#&gt; [111] scales_1.3.0                lmtest_0.9-40              </span></span>
<span><span class="co">#&gt; [113] rappdirs_0.3.3              stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [115] digest_0.6.37               goftest_1.2-3              </span></span>
<span><span class="co">#&gt; [117] spatstat.utils_3.1-0        rmarkdown_2.28             </span></span>
<span><span class="co">#&gt; [119] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [121] pkgconfig_2.0.3             sparseMatrixStats_1.16.0   </span></span>
<span><span class="co">#&gt; [123] MatrixGenerics_1.16.0       highr_0.11                 </span></span>
<span><span class="co">#&gt; [125] dbplyr_2.5.0                fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [127] rlang_1.1.4                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [129] ggthemes_5.1.0              UCSC.utils_1.0.0           </span></span>
<span><span class="co">#&gt; [131] shiny_1.9.1                 DelayedMatrixStats_1.26.0  </span></span>
<span><span class="co">#&gt; [133] farver_2.1.2                jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [135] zoo_1.8-12                  jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt; [137] BiocParallel_1.38.0         mclust_6.1.1               </span></span>
<span><span class="co">#&gt; [139] BiocSingular_1.20.0         magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [141] Formula_1.2-5               scuttle_1.14.0             </span></span>
<span><span class="co">#&gt; [143] GenomeInfoDbData_1.2.12     dotCall64_1.1-1            </span></span>
<span><span class="co">#&gt; [145] patchwork_1.3.0             munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [147] Rcpp_1.0.13                 viridis_0.6.5              </span></span>
<span><span class="co">#&gt; [149] reticulate_1.39.0           stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [151] zlibbioc_1.50.0             MASS_7.3-60.2              </span></span>
<span><span class="co">#&gt; [153] org.Hs.eg.db_3.19.1         plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [155] parallel_4.4.1              PRECAST_1.6.5              </span></span>
<span><span class="co">#&gt; [157] listenv_0.9.1               ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt; [159] deldir_2.0-4                Biostrings_2.72.1          </span></span>
<span><span class="co">#&gt; [161] splines_4.4.1               tensor_1.5                 </span></span>
<span><span class="co">#&gt; [163] hms_1.1.3                   igraph_2.0.3               </span></span>
<span><span class="co">#&gt; [165] ggpubr_0.6.0                spatstat.geom_3.3-3        </span></span>
<span><span class="co">#&gt; [167] ggsignif_0.6.4              RcppHNSW_0.6.0             </span></span>
<span><span class="co">#&gt; [169] reshape2_1.4.4              biomaRt_2.60.1             </span></span>
<span><span class="co">#&gt; [171] stats4_4.4.1                ScaledMatrix_1.12.0        </span></span>
<span><span class="co">#&gt; [173] evaluate_1.0.0              httpuv_1.6.15              </span></span>
<span><span class="co">#&gt; [175] RANN_2.6.2                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt; [177] polyclip_1.10-7             scattermore_1.2            </span></span>
<span><span class="co">#&gt; [179] ggplot2_3.5.1               rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt; [181] broom_1.0.7                 xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [183] RSpectra_0.16-2             rstatix_0.7.2              </span></span>
<span><span class="co">#&gt; [185] later_1.3.2                 viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [187] ragg_1.3.3                  tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [189] memoise_2.0.1               beeswarm_0.4.0             </span></span>
<span><span class="co">#&gt; [191] AnnotationDbi_1.68.0        IRanges_2.38.1             </span></span>
<span><span class="co">#&gt; [193] cluster_2.1.6               globals_0.16.3</span></span></code></pre></div>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Xiao Zhang, Yi Yang, Peng Xie, Chengqi Lin, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
